crates/extensions/c8y_mapper_ext/src/tests.rs

```diff
@@ -1069,6 +1069,51 @@ async fn mapper_publishes_supported_operations_for_child_device() {
     .await;
 }
 
+#[tokio::test]
+async fn mapping_child_device_dirs_with_forbidden_characters() {
+    // The test assures tedge-mapper checks if there is a directory for operations for child devices, then it reads and
+    // correctly publishes supported operations message for that child on to `c8y/s/us/child1`
+    // and verifies that message.
+    let cfg_dir = TempTedgeDir::new();
+    create_thin_edge_child_operations(&cfg_dir, "my#complex+child2", vec!["c8y_ChildTestOp1"]);
+
+    let (mqtt, _http, _fs, _timer) = spawn_c8y_mapper_actor(&cfg_dir, false).await;
+    let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+    mqtt.skip(6).await; //Skip all init messages
```

Comment:
suggestion: IMO doing this results in a test not only asserting that "mapper correctly creates child devices from directory names with invalid characters", but also asserting that "there are always 6 init messages", and adding or removing a message in a future will cause a test to fail with a trivial fix, but the annoying thing is the number of the tests that assume number of init messages, and it will have to be changed in every one of them. IIRC I once had to do this, and it also unnecessarily increases the number of files changed, so I'd prefer if we did not assert things like "number of init messages" in a test which shouldn't really care about that.

Reply:
> so I'd prefer if we did not assert things like "number of init messages" in a test which shouldn't really care about that.

Yeah, the `skip` was a way of avoiding asserting all the init messages in each and every test. But, as you highlighted, if the number of `init` messages change, we'll end up updating every test. So, to avoid this, I can move this skip to a dedicated function and have all the tests use that function. That way, in future if we add more init messages, just that single function needs to be updated.

Reply:
> I can move this skip to a dedicated function and have all the tests use that function. That way, in future if we add more init messages, just that single function needs to be updated.

Extracting `6` to a `INIT_MESSAGES_NUMBER` constant will have the same effect and will be a bit more transparent. What I rather meant was to write the test in a way as to not refer to any other messages and only refer to messages we're interested about.

As in the test we're spinning up all of these actors, and we get input and output channels with messages in them. In the test we check if certain input messages when processed by the actor produce certain output messages, so we can just check that these messages we expect are inside the output channel, without expecting to read any other messages that are not part of the test. Then we don't have to even care how many init messages there are.

I don't know how feasible it is currently, as `assert_received_contains_str` function is probably actively reading from the oldest message and taking messages off the channel, so right now it can stay as it is, just wanted to present my viewpoint and ask if you agree in principle.

Reply:
> Extracting 6 to a INIT_MESSAGES_NUMBER constant will have the same effect and will be a bit more transparent.

Having a dedicated function that'll just exhaust the init messages will help in doing additional validations as well, rather than just skipping a fixed number of messages.

> so we can just check that these messages we expect are inside the output channel, without expecting to read any other messages that are not part of the test

This is the tricky bit. Sometimes we end up sending the same messages during the `init` as well as the result of a mapping. Examples: child device creation messages(`101,...`), supported operation messages(`118,...`) etc. So, while doing something like you proposed where we just search for the desired message in a stream, we need to make sure that we're not hitting the wrong one.

> as assert_received_contains_str function is probably actively reading from the oldest message and taking messages off the channel, so right now it can stay as it is, just wanted to present my viewpoint and ask if you agree in principle.

Yes, that's the case and if we can write a similar helper that can search for the desired pattern ignoring false positives, that'd be great.